{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a rel=\"license\" href=\"http://creativecommons.org/licenses/by/4.0/\"><img alt=\"Creative Commons License\" style=\"border-width:0\" src=\"https://i.creativecommons.org/l/by/4.0/88x31.png\" /></a><br /><span xmlns:dct=\"http://purl.org/dc/terms/\" property=\"dct:title\"><b>Computational Optimal Transport</b></span> by <a xmlns:cc=\"http://creativecommons.org/ns#\" href=\"http://mate.unipv.it/gualandi\" property=\"cc:attributionName\" rel=\"cc:attributionURL\">Stefano Gualandi</a> is licensed under a <a rel=\"license\" href=\"http://creativecommons.org/licenses/by/4.0/\">Creative Commons Attribution 4.0 International License</a>. Based on a project at <a xmlns:dct=\"http://purl.org/dc/terms/\" href=\"https://github.com/mathcoding/opt4ds\" rel=\"dct:source\">https://github.com/mathcoding/compopt</a>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computing Wasserstein Barycenters\n",
    "In this book, we discuss the fundamental challenges in computing Wasserstein Barycenters.\n",
    "\n",
    "The exercise are based on the following references:\n",
    "\n",
    "* Chapter 9 in [Computational Optimal Transport]()\n",
    "* Benamou, J. D., Carlier, G., Cuturi, M., Nenna, L., & Peyr√©, G. (2015). *Iterative Bregman projections for regularized transportation problems*. SIAM Journal on Scientific Computing, 37(2), A1111-A1138.\n",
    "* Bouchet, P.Y., Gualandi, S. and Rousseau, L.M., 2020. *Primal heuristics for wasserstein barycenters*. CPAIOR 2020, Proceedings 17 (pp. 239-255), Springer.\n",
    "* Auricchio, G., Bassetti, F., Gualandi, S. and Veneroni, M., 2019. *Computing Wasserstein barycenters via linear programming*. CPAIOR 2019 Proceedings 16 (pp. 355-363), Springer.\n",
    "\n",
    "**IMPORTANT:** One of the best application of OT is in the study of gene expression profiles:\n",
    "\n",
    "* Schiebinger, G., Shu, J., Tabaka, M., Cleary, B., Subramanian, V., Solomon, A., Gould, J., Liu, S., Lin, S., Berube, P. and Lee, L., 2019. [Optimal-transport analysis of single-cell gene expression identifies developmental trajectories in reprogramming](https://www.cell.com/cell/pdf/S0092-8674(19)30039-X.pdf). Cell, 176(4), pp.928-943.\n",
    "\n",
    "We suggest to watch the following YouTube research seminars:\n",
    "\n",
    "* [Statistical and Computational aspects of Wasserstein Barycenters - P. Rigollet @ MAD+ (8 Apr 2020)](https://www.youtube.com/watch?v=MMlR-PeMsgg)\n",
    "* [Pavel Dvurechensky - \"Wasserstein barycenters from the computational perspective\" | MoCCA'20](https://www.youtube.com/watch?v=j93j9W4JgyI&t=245s)\n",
    "* [Darina Dvinskikh - \"Decentralized Algorithms for Wasserstein Barycenters\" | MoCCA'20](https://www.youtube.com/watch?v=cC9xGHSx3B0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wasserstein Barycenters in 1D\n",
    "The following two exercises are from [POT examples](https://pythonot.github.io/auto_examples/index.html#wasserstein-barycenters).\n",
    "\n",
    "### Using the optimal transportation plan for interpolation between two measures\n",
    "Try to play with the following code, by changing the value of *aplha* (the weights of the barycenter)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import ot\n",
    "# necessary for 3d plot even if not used\n",
    "from mpl_toolkits.mplot3d import Axes3D  # noqa\n",
    "from matplotlib.collections import PolyCollection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# nb bins\n",
    "n = 100\n",
    "\n",
    "# bin positions\n",
    "x = np.arange(n, dtype=np.float64)\n",
    "\n",
    "# Gaussian distributions\n",
    "a1 = ot.datasets.make_1D_gauss(n, m=20, s=5)  # m= mean, s= std\n",
    "a2 = ot.datasets.make_1D_gauss(n, m=60, s=8)\n",
    "\n",
    "# creating matrix A containing all distributions\n",
    "A = np.vstack((a1, a2)).T\n",
    "n_distributions = A.shape[1]\n",
    "\n",
    "# loss matrix + normalization\n",
    "M = ot.utils.dist0(n)\n",
    "M /= M.max()\n",
    "\n",
    "print(M.shape)\n",
    "\n",
    "alpha = 0.5  # 0<=alpha<=1\n",
    "weights = np.array([1 - alpha, alpha])\n",
    "\n",
    "# l2bary\n",
    "bary_l2 = A.dot(weights)\n",
    "\n",
    "# wasserstein\n",
    "reg = 1e-3\n",
    "bary_wass = ot.bregman.barycenter(A, M, reg, weights)\n",
    "\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, tight_layout=True, num=1)\n",
    "ax1.plot(x, A, color=\"black\")\n",
    "ax1.set_title('Distributions')\n",
    "\n",
    "ax2.plot(x, bary_l2, 'r', label='l2')\n",
    "ax2.plot(x, bary_wass, 'g', label='Wasserstein')\n",
    "ax2.set_title('Barycenters')\n",
    "\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To automate the change of alpha, and having a plot in 3D (of 1D distributions) try the following code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try 11 values of alpha between 0 and 1\n",
    "n_alpha = 11\n",
    "alpha_list = np.linspace(0, 1, n_alpha)\n",
    "\n",
    "B_l2 = np.zeros((n, n_alpha))\n",
    "\n",
    "B_wass = np.copy(B_l2)\n",
    "\n",
    "# Compute the barycenter for each value of alpha\n",
    "for i in range(n_alpha):\n",
    "    alpha = alpha_list[i]\n",
    "    weights = np.array([1 - alpha, alpha])\n",
    "    B_l2[:, i] = A.dot(weights)\n",
    "    B_wass[:, i] = ot.bregman.barycenter(A, M, reg, weights)\n",
    "\n",
    "plt.figure(1)\n",
    "\n",
    "# Plot the barycenter interpolations\n",
    "cmap = plt.cm.get_cmap('viridis')\n",
    "verts = []\n",
    "zs = alpha_list\n",
    "for i, z in enumerate(zs):\n",
    "    ys = B_l2[:, i]\n",
    "    verts.append(list(zip(x, ys)))\n",
    "\n",
    "ax = plt.gcf().add_subplot(projection='3d')\n",
    "\n",
    "poly = PolyCollection(verts, facecolors=[cmap(a) for a in alpha_list])\n",
    "poly.set_alpha(0.7)\n",
    "ax.add_collection3d(poly, zs=zs, zdir='y')\n",
    "ax.set_xlabel('x')\n",
    "ax.set_xlim3d(0, n)\n",
    "ax.set_ylabel('$\\\\alpha$')\n",
    "ax.set_ylim3d(0, 1)\n",
    "ax.set_zlabel('')\n",
    "ax.set_zlim3d(0, B_l2.max() * 1.01)\n",
    "plt.title('Barycenter interpolation with l2')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, try to compute the Wasserstein Barycenters with the following code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(1)\n",
    "cmap = plt.cm.get_cmap('viridis')\n",
    "verts = []\n",
    "zs = alpha_list\n",
    "for i, z in enumerate(zs):\n",
    "    ys = B_wass[:, i]\n",
    "    verts.append(list(zip(x, ys)))\n",
    "\n",
    "ax = plt.gcf().add_subplot(projection='3d')\n",
    "\n",
    "poly = PolyCollection(verts, facecolors=[cmap(a) for a in alpha_list])\n",
    "poly.set_alpha(0.7)\n",
    "ax.add_collection3d(poly, zs=zs, zdir='y')\n",
    "ax.set_xlabel('x')\n",
    "ax.set_xlim3d(0, n)\n",
    "ax.set_ylabel('$\\\\alpha$')\n",
    "ax.set_ylim3d(0, 1)\n",
    "ax.set_zlabel('')\n",
    "ax.set_zlim3d(0, B_l2.max() * 1.01)\n",
    "plt.title('Barycenter interpolation with Wasserstein')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare Barycenters computed with Exact and Entropic Optimal Transport\n",
    "The following code show the difference in practice for computing Barycenters of using the exact Optimal Transport solved with a simplex algorithm, versus the use of an entropic regularized solver."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pylab as pl\n",
    "import ot\n",
    "# necessary for 3d plot even if not used\n",
    "from mpl_toolkits.mplot3d import Axes3D  # noqa\n",
    "from matplotlib.collections import PolyCollection  # noqa\n",
    "\n",
    "problems = []\n",
    "\n",
    "n = 100  # nb bins\n",
    "\n",
    "# bin positions\n",
    "x = np.arange(n, dtype=np.float64)\n",
    "\n",
    "# Gaussian distributions\n",
    "# Gaussian distributions\n",
    "a1 = ot.datasets.make_1D_gauss(n, m=20, s=5)  # m= mean, s= std\n",
    "a2 = ot.datasets.make_1D_gauss(n, m=60, s=8)\n",
    "\n",
    "# creating matrix A containing all distributions\n",
    "A = np.vstack((a1, a2)).T\n",
    "n_distributions = A.shape[1]\n",
    "\n",
    "# loss matrix + normalization\n",
    "M = ot.utils.dist0(n)\n",
    "M /= M.max()\n",
    "\n",
    "alpha = 0.5  # 0<=alpha<=1\n",
    "weights = np.array([1 - alpha, alpha])\n",
    "\n",
    "# l2bary\n",
    "bary_l2 = A.dot(weights)\n",
    "\n",
    "# wasserstein\n",
    "reg = 1e-3\n",
    "ot.tic()\n",
    "bary_wass = ot.bregman.barycenter(A, M, reg, weights)\n",
    "ot.toc()\n",
    "\n",
    "ot.tic()\n",
    "bary_wass2 = ot.lp.barycenter(A, M, weights)\n",
    "ot.toc()\n",
    "\n",
    "problems.append([A, [bary_l2, bary_wass, bary_wass2]])\n",
    "\n",
    "# STAIRS DATA\n",
    "a1 = 1.0 * (x > 10) * (x < 50)\n",
    "a2 = 1.0 * (x > 60) * (x < 80)\n",
    "\n",
    "a1 /= a1.sum()\n",
    "a2 /= a2.sum()\n",
    "\n",
    "# creating matrix A containing all distributions\n",
    "A = np.vstack((a1, a2)).T\n",
    "n_distributions = A.shape[1]\n",
    "\n",
    "# loss matrix + normalization\n",
    "M = ot.utils.dist0(n)\n",
    "M /= M.max()\n",
    "\n",
    "# l2bary\n",
    "bary_l2 = A.dot(weights)\n",
    "\n",
    "# wasserstein\n",
    "reg = 1e-3\n",
    "ot.tic()\n",
    "bary_wass = ot.bregman.barycenter(A, M, reg, weights)\n",
    "ot.toc()\n",
    "\n",
    "ot.tic()\n",
    "bary_wass2 = ot.lp.barycenter(A, M, weights)\n",
    "ot.toc()\n",
    "\n",
    "problems.append([A, [bary_l2, bary_wass, bary_wass2]])\n",
    "\n",
    "# DIRAC DELTA DATA\n",
    "a1 = np.zeros(n)\n",
    "a2 = np.zeros(n)\n",
    "\n",
    "a1[10] = .25\n",
    "a1[20] = .5\n",
    "a1[30] = .25\n",
    "a2[80] = 1\n",
    "\n",
    "a1 /= a1.sum()\n",
    "a2 /= a2.sum()\n",
    "\n",
    "# creating matrix A containing all distributions\n",
    "A = np.vstack((a1, a2)).T\n",
    "n_distributions = A.shape[1]\n",
    "\n",
    "# loss matrix + normalization\n",
    "M = ot.utils.dist0(n)\n",
    "M /= M.max()\n",
    "\n",
    "# l2bary\n",
    "bary_l2 = A.dot(weights)\n",
    "\n",
    "# wasserstein\n",
    "reg = 1e-3\n",
    "ot.tic()\n",
    "bary_wass = ot.bregman.barycenter(A, M, reg, weights)\n",
    "ot.toc()\n",
    "\n",
    "ot.tic()\n",
    "bary_wass2 = ot.lp.barycenter(A, M, weights)\n",
    "ot.toc()\n",
    "\n",
    "problems.append([A, [bary_l2, bary_wass, bary_wass2]])\n",
    "\n",
    "pl.figure(1, (20, 6))\n",
    "pl.clf()\n",
    "\n",
    "for i in range(nbm):\n",
    "    A = problems[i][0]\n",
    "    bary_l2 = problems[i][1][0]\n",
    "    bary_wass = problems[i][1][1]\n",
    "    bary_wass2 = problems[i][1][2]\n",
    "\n",
    "    pl.subplot(2, nbm, 1 + i)\n",
    "    for j in range(n_distributions):\n",
    "        pl.plot(x, A[:, j])\n",
    "    if i == nbm2:\n",
    "        pl.title('Distributions')\n",
    "    pl.xticks(())\n",
    "    pl.yticks(())\n",
    "\n",
    "    pl.subplot(2, nbm, 1 + i + nbm)\n",
    "\n",
    "    pl.plot(x, bary_l2, 'r', label='L2 (Euclidean)')\n",
    "    pl.plot(x, bary_wass, 'g', label='Reg Wasserstein')\n",
    "    pl.plot(x, bary_wass2, 'b', label='LP Wasserstein')\n",
    "    if i == nbm - 1:\n",
    "        pl.legend()\n",
    "    if i == nbm2:\n",
    "        pl.title('Barycenters')\n",
    "\n",
    "    pl.xticks(())\n",
    "    pl.yticks(())\n",
    "    \n",
    "pl.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Implementing and testing the Iterative Bregman Projection algorithm\n",
    "Consider the 10 double-centric ellipses, loaded from a Matlab .dat file with the following snippet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "\n",
    "def ReadEggs():\n",
    "    imported_ellipses = scipy.io.loadmat('../data/ellipses.mat')\n",
    "    ellipses=imported_ellipses['ellipses']\n",
    "    return ellipses\n",
    "\n",
    "def PlotEggs(ellipses):\n",
    "    fig, axs = plt.subplots(2, 5, figsize=(20, 8))  # 2 rows, 5 columns\n",
    "\n",
    "    for i in range(2):  # Rows\n",
    "        for j in range(5):  # Columns\n",
    "            axs[i, j].imshow(ellipses[i*5+j], cmap='binary')  # Example plot\n",
    "            axs[i, j].set_xticks([])  # Remove x-axis tick labels\n",
    "            axs[i, j].set_yticks([])  # Remove y-axis tick labels#axs[i, j].set_title(f'Plot {(i*5)+j+1}')\n",
    "\n",
    "    plt.tight_layout()  # Adjust subplots to fit into the figure area.\n",
    "    plt.show()  # Display the figure\n",
    "\n",
    "ellipses = ReadEggs()\n",
    "PlotEggs(ellipses)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**QUESTION:** Which is the empirical distribution that best represent this 10 samples?\n",
    "\n",
    "You can try using the POT library using either the function [convolutional barycenter 2d](https://pythonot.github.io/gen_modules/ot.bregman.html#ot.bregman.convolutional_barycenter2d) or the [convolutional barycenter debiased](https://pythonot.github.io/gen_modules/ot.bregman.html#ot.bregman.convolutional_barycenter2d_debiased) as in the following script (try separately the two functions, using the debiased as second option):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ConvBary(A, reg=0.004):\n",
    "    fig, axs = plt.subplots(3, 5, figsize=(20, 8))  # 2 rows, 5 columns\n",
    "    for i in range(2):  # Rows\n",
    "        for j in range(5):  # Columns\n",
    "            axs[i, j].imshow(ellipses[i*5+j], cmap='binary')  # Example plot\n",
    "            axs[i, j].set_xticks([])  # Remove x-axis tick labels\n",
    "            axs[i, j].set_yticks([])  # Remove y-axis tick labels#axs[i, j].set_title(f'Plot {(i*5)+j+1}')\n",
    "\n",
    "    for j, r in enumerate([0.1, 0.01, 0.005, 0.002, 0.001]):\n",
    "        \n",
    "        B = ot.bregman.convolutional_barycenter2d(A, r)\n",
    "        \n",
    "        # COMMENT THE PREVIOUS LINE AND UNCOMMENT THE FOLLOWING LINE TO USE THE NON-DEBIASED VERSION\n",
    "        #B = ot.bregman.convolutional_barycenter2d_debiased(A, r)\n",
    "\n",
    "        axs[2, j].imshow(B, cmap='binary')  # Example plot\n",
    "        axs[2, j].set_xticks([])  # Remove x-axis tick labels\n",
    "        axs[2, j].set_yticks([])  # Remove y-axis tick labels\n",
    "        axs[2, j].set_title(f'reg={r}')\n",
    "\n",
    "    plt.tight_layout()  # Adjust subplots to fit into the figure area.\n",
    "    plt.show()\n",
    "\n",
    "ConvBary(ellipses)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Implement your own Entropic Wasserstein Barycenter algorithm\n",
    "Following the suggestions given during the lecture, implement your own algorithm to compute a Wasserstein Barycenter.\n",
    "\n",
    "Possible options:\n",
    "\n",
    "1. Implement the Iterative Bregman Projection as explained during the class (but exploiting the 2D grid structure of the cost matrix)\n",
    "2. Implement any of the iterative pairing heuristic as explained in the slides\n",
    "3. **IMPORTANT:** Design your own new algorithm, make extensive computational test providing computational evidence of your algorithm, and write a paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def WB(A, C, reg=0.05):\n",
    "    # Prepare cost matrix\n",
    "    l = A.shape[0]\n",
    "    n = A.shape[1]\n",
    "    N = n**2\n",
    "\n",
    "    # TODO: complete with your code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to save time, you can precompute once the cose matrix of images 60x60, save it as a binary file, and loaded when ncessary.\n",
    "\n",
    "Look at following code snippet."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GenerateCostMatrix(n, p=2):\n",
    "    N = n**2\n",
    "    C = np.zeros((N, N))\n",
    "    for i in range(n):\n",
    "        for j in range(n):\n",
    "            for v in range(n):\n",
    "                for w in range(n):\n",
    "                    if p == 1:\n",
    "                        C[i*n+j, v*n+w] = np.abs(i-v) + np.abs(j-w)\n",
    "                    elif p == 2:\n",
    "                        C[i*n+j, v*n+w] = (i-v)**2 + (j-w)**2\n",
    "    return C\n",
    "\n",
    "if False:\n",
    "    C = GenerateCostMatrix(60, p=2)\n",
    "    np.save('n60p2s.npy', C)\n",
    "else:\n",
    "    C = np.load('../data/n60p2.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
